dallinger/prolific.py

```diff
@@ -281,3 +281,91 @@ def _req(self, method: str, endpoint: str, **kw) -> dict:
             raise ProlificServiceException(json.dumps(error))
 
         return parsed
+
+
+class DevProlificService(ProlificService):
+    """Wrapper that mocks the Prolific REST API and instead of making requests it writes to the log."""
+
+    def __init__(self, api_token: str, api_version: str, referer_header: str):
+        super().__init__(api_token, api_version, referer_header)
+
+    def approve_participant_session(self, session_id: str) -> dict:
+        self._req(
+            method="POST",
+            endpoint=f"/submissions/{session_id}/transition/",
+            json={"action": "APPROVE"},
+        )
+        return True
+
+    def draft_study(
```

Comment:
I think we need to brainstorm the strategy a bit more, as I think there is too much copy-pasting in this `draft_study` function. The one piece of logic you are changing is that you want to patch the response from `_req`; that should not necessitate copy-pasting the whole function. 

I propose the following. Do not override `draft_study` at all. Instead, override `_req` like this:

```py
def _req(self, method: str, endpoint: str, **kw) -> dict:
    logger.info("Simulated Prolific API call: method = {}, endpoint = {}, kwarg = {}"...)

    if endpoint == "/studies/":
        return {"id": "prolific-user-id", "external_study_url": "external-study-url"}
    elif endpoint.startswith("/bulk-bonus-payments/")
        ...
```

Similar logic can be applied to the other methods. Do you see what I mean?

Reply:
Followed your proposal. :+1: